<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [9]:
import pandas as pd
import psycopg2
import plotly.express as px
import requests 
import urllib.parse as req

In [11]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [12]:
# текст запроса
query_3_1 = f'''SELECT COUNT(id)
                FROM public.vacancies
'''
result = pd.read_sql_query(query_3_1,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [13]:
# результат запроса
result

count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [14]:
# текст запроса
query_3_2 = f'''SELECT COUNT(id)
                FROM public.employers
'''
result = pd.read_sql_query(query_3_2,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [15]:
# результат запроса
result

count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [16]:
# текст запроса
query_3_3 = f'''SELECT COUNT(id)
                FROM public.areas
'''
result = pd.read_sql_query(query_3_3,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [17]:
# результат запроса
result


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [18]:
# текст запроса
query_3_4 = f'''SELECT COUNT(id)
                FROM public.industries
'''
result = pd.read_sql_query(query_3_4,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [19]:
# результат запроса
result

count
0    294

***

# Выводы предварительного анализа данных:

* В нашей базе данных весьма обширное кол-во вакансий - около 50 тысяч, от почти 24 тысяч рабтодателей, это означает, сухой оценке, 
каждый работодатель нуждается по меньшей мере в 2 новых работниках.
*  Количество регионов насчитывает почти 1400 - что является обширной территоррией, благодаря чему, люди из различных регионов не ограничены в выборе вакансий. И на последок, вакансии в нашей базе данных предлагают работу почти  в 300 различных сферах деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [20]:
# текст запроса
query_4_1 = f'''
SELECT 
    a.name area,
    COUNT(v.id) cnt
FROM public.vacancies v
    JOIN public.areas a ON v.area_id = a.id
GROUP BY 1
HAVING COUNT(v.id)>1000
ORDER BY 2 DESC
LIMIT 12
'''
df_4_1 = pd.read_sql_query(query_4_1,connection)
df_4_1

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area   cnt
0            Москва  5333
1   Санкт-Петербург  2851
2             Минск  2112
3       Новосибирск  2006
4            Алматы  1892
5      Екатеринбург  1698
6   Нижний Новгород  1670
7            Казань  1415
8         Краснодар  1301
9            Самара  1144
10   Ростов-на-Дону  1131
11          Воронеж  1063

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [21]:
# текст запроса
query_4_2 = f'''
select 
    count(*)               
from vacancies
where (salary_from is not null or salary_to is not null)
'''
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [22]:
# текст запроса
query_4_3 = f'''
SELECT 
    AVG(salary_from)
    FROM public.vacancies 
UNION
SELECT
    AVG(salary_to)
    FROM public.vacancies  
'''      
df_4_3 = pd.read_sql_query(query_4_3, connection)
round(df_4_3)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


avg
0   71065.0
1  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [23]:
# текст запроса
query_4_4 = f'''
SELECT  
    COUNT(*),
    v.schedule,
    v.employment
FROM public.vacancies v
GROUP BY 2,3
ORDER BY COUNT(*) DESC
OFFSET 1 LIMIT 1'''
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count          schedule        employment
0   7802  Удаленная работа  Полная занятость

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [24]:
# текст запроса
query_4_5 = f'''
SELECT
    COUNT(*),
    experience
FROM public.vacancies
GROUP BY 2
ORDER BY 1 ASC'''
df_4_5 = pd.read_sql_query(query_4_5,connection)
df_4_5

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

***

# Выводы по детальному анализу данных:

* В детальном анализе данных мы выявили города в которых вакансий больше всего: МСК, СПБ, Минск, Новосибирск, Алматы. Результат вполне ожидаемый - в крупных гороах больше всего вакансий. 
* Далее мы рассмотрели зарплаты, то-есть у какого кол-ва вакансий заполнено хотя бы одно из двух полей - нижняя и вверхняя грань З/П - и выявили, что у почти половины вакансий (24 073) заполнено хотя-бы одно из двух полей. 
* Потом мы выявили среднее значение этих двух полей - 71065 тысячи рублей для нижней грани и 110537 тыс. для верхней.  
* Так же мы определили самые популярные сочетания типов рабочего графика и типов трудоустройства в наших вакансиях. Второе место из этих сочитаний является "удаленная работы - полная занятость" - то к чему в целом многие стремятся. 
* И на последок мы обнаружили самый востребованный опыт работы среди представленных вакансий, самым популярным оказался "от 1 года до 3 лет", так как это соответсвует начальному уровню опыта работы или позиции "Джуниор", которых используют в качестве более дешевой рабочей силы, для выполнения не сложных задач.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [25]:
# текст запроса
query_5_1 = f'''
(SELECT 
    e.name,
    COUNT(v.id)
FROM public.employers e
    JOIN vacancies v ON v.employer_id = e.id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1)
UNION
(SELECT 
    e.name,
    COUNT(v.id)
FROM public.employers e
    JOIN vacancies v ON v.employer_id = e.id
GROUP BY 1
ORDER BY 2 DESC
OFFSET 4 LIMIT 1)
'''
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0  Газпром нефть    331
1         Яндекс   1933

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
# текст запроса
query_5_2 = f'''
SELECT  
    a.name,
    COUNT(DISTINCT(e.id)),
    COUNT(v.id)
FROM public.areas a
    JOIN public.employers e ON a.id = e.area
    LEFT JOIN public.vacancies v ON a.id = v.area_id
WHERE v.id IS NULL
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1
'''
df_5_2 = pd.read_sql_query(query_5_2,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [27]:
# результат запроса
df_5_2

name  count  count
0  Россия    410      0

In [28]:
# Vacancies from different areas

# importing needed packages
# from urllib.request import urlopen
# import json

# # Opening a GeoJson file from github:
# with urlopen("https://raw.githubusercontent.com/wmgeolab/geoBoundaries/042926b1255dac674bf2ce92df9cbb97432a891d/releaseData/gbOpen/RUS/ADM0/geoBoundaries-RUS-ADM0_simplified.geojson") as response:
#     russia_regions = json.load(response)
    
# query = f'''
# SELECT  
#     a.name,
#     COUNT(DISTINCT(e.id)) employers,
#     COUNT(v.id) vacancies
# FROM public.areas a
#     JOIN public.employers e ON a.id = e.area
#     LEFT JOIN public.vacancies v ON a.id = v.area_id
# GROUP BY 1
# ORDER BY 2 DESC
# '''
# df = pd.read_sql_query(query,connection)  
    
# fig = px.choropleth(
#     df,
#     geojson=russia_regions,
#     color='vacancies',
#     scope='asia'
# )
# fig.show()


К сожалению у меня не запустился код, выше, вышел бы весьма наглядный и хороший график, возможно я что-то сделал не так.

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [29]:
# текст запроса
query_5_3 = f'''
SELECT
    e.name,
    COUNT(DISTINCT(a.name))
FROM employers e
    JOIN vacancies v ON v.employer_id = e.id
    JOIN areas a ON v.area_id = a.id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1
'''
df_5_3 = pd.read_sql_query(query_5_3,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [30]:
# результат запроса
df_5_3

name  count
0  Яндекс    181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [31]:
# текст запроса
query_5_4 = f'''
SELECT
    COUNT(e.id)
FROM public.employers e
    LEFT JOIN public.employers_industries i ON i.employer_id = e.id
WHERE i.industry_id IS NULL
'''
df_5_4 = pd.read_sql_query(query_5_4,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [32]:
# результат запроса
df_5_4

count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [33]:
# текст запроса
query_5_5 = f''' 
SELECT
    e.name
FROM public.employers e
    JOIN public.employers_industries i ON i.employer_id = e.id
GROUP BY 1
HAVING COUNT(i.industry_id)=4
ORDER BY 1 ASC
OFFSET 2 LIMIT 1
'''
df_5_5 = pd.read_sql_query(query_5_5, connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [34]:
# результат запроса
df_5_5

name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [35]:
# текст запроса
query_5_6 = f'''  
SELECT
    COUNT(e.id)
FROM public.employers e
    JOIN public.employers_industries ie ON ie.employer_id = e.id
    JOIN public.industries i ON i.id = ie.industry_id
WHERE i.name = 'Разработка программного обеспечения'
'''
df_5_6 = pd.read_sql_query(query_5_6,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [36]:
# результат запроса
df_5_6

count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [37]:
# код для получения списка городов-милионников
#creating url
path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

table_1 = pd.read_html(url)

df = table_1[1]

cities_list = df['Город'].to_list()
cities_list[13] = 'Воронеж'     #извините за такой колхоз, но ментор не отвечал как это исправить)
cities = tuple(cities_list)
cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [38]:
query_5_7 = f'''  
SELECT
    a.name,
    count(v.id) cnt
FROM public.vacancies v
    LEFT JOIN public.areas a ON a.id = v.area_id
    LEFT JOIN public.employers e ON v.employer_id = e.id
WHERE e.name = 'Яндекс'
    AND a.name in {cities}
GROUP BY 1

UNION ALL

SELECT 
    'Total', count(v.id)
FROM public.vacancies v
    LEFT JOIN public.areas a ON a.id = v.area_id
    LEFT JOIN public.employers e ON v.employer_id = e.id
WHERE e.name = 'Яндекс'
    AND a.name in {cities}
ORDER BY 2
'''
df_5_7 = pd.read_sql_query(query_5_7, connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [39]:
# результат запроса
df_5_7.shape[0]
df_5_7

name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

# Выводы по анализу работодателей:

* Для начала, мы решили узнать, какие работодатели лидируют по количеству вакансий, из нашего запроса мы выявили, что Яндекс - как и ожидалось - занимает первую позицию по кол-ву вакансий, а пятерку лидиров
же занимает "Газпром нефть". 
* Далее мы нашли регионы в которых нет вакансий, но при этом наибольшее количество работодателей. Наш запрос выдал Россию, очевидно, это не совсем так, в России полно вакансий,
единственное в самой вакансии указывается не страна, а контретный регион, например Московская область, Москва, Воронеж и тд. и тп. А вот работодатели, или же компании, в большенстве указывают свой регион как "Россия" оттого и такой ответ. 
* Затем мы считали количество регионов для каждого работодателя, в которых он публикует свои вакансии, где опредили вновь лидерство Яндекса с 181 регионом. 
* Потом мы посчитали кол-во работодателей у которых не указана сфера деятельности в вакансии, их оказалось 8419. Среди компаний, у которых указано 4 сферы деятельности в вакансии, на третьем месте в отсортированной по алфавиту таблице оказалась компания "2ГИС". 
* Так-же мы узнали у какого количества компаний в сфере деятельности указана "Разработка программного обеспечения", их оказалось 3553. 
* И напоследок мы решили для компании "Яндекс", найти количество вакансий в регионах-миллионниках. В результате запроса мы выявили, что всего компания "Яндекс" публекует вакансии в 16 регионах-миллионниках и что на первом месте по кол-ву вакансий находится Москва.

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
# текст запроса
query_6_1 = f'''  
SELECT COUNT(v.id)
FROM public.vacancies v
WHERE LOWER(v.name) LIKE '%%data%%' OR LOWER(v.name) LIKE '%%данн%%'
'''
df_6_1 = pd.read_sql_query(query_6_1,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [41]:
# результат запроса
df_6_1

count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [42]:
DS_query = "(lower(name) LIKE '%data scientist%' OR \
        lower(name) LIKE '%data science%' OR\
        lower(name) LIKE '%исследователь данных%' OR\
        (name LIKE '%ML%' AND name NOT LIKE '%HTML%') OR\
        lower(name) LIKE '%machine learning%' OR \
        lower(name) LIKE '%машинн%обучен%')"

In [43]:
# текст запроса
query_6_2 = f'''
SELECT
     COUNT(*) 
FROM vacancies v
WHERE {DS_query}
     AND (name ilike '%junior%' OR experience ilike '%нет опыта%' OR employment ilike '%стажировка%')
'''
df_6_2 = pd.read_sql_query(query_6_2,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [44]:
# результат запроса
df_6_2

count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [45]:
# текст запроса
query_6_3 = f'''
SELECT
     COUNT(*) 
FROM vacancies 
WHERE {DS_query}
     
     AND (lower(key_skills) LIKE '%sql%' OR lower(key_skills) LIKE '%postgres%')
'''
df_6_3 = pd.read_sql_query(query_6_3,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [46]:
# результат запроса
df_6_3

count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [47]:
# текст запроса
query_6_4 = f'''
SELECT
     COUNT(*) 
FROM vacancies 
WHERE {DS_query}
     
     AND lower(key_skills) LIKE '%python%'
'''
df_6_4 = pd.read_sql_query(query_6_4,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [48]:
# результат запроса
df_6_4

count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [49]:
# текст запроса
query_6_5 = f'''  
SELECT
    ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills,Chr(9),'')))+1,2) avg
FROM public.vacancies
WHERE {DS_query}
    AND key_skills IS NOT NULL
'''
df_6_5 = pd.read_sql_query(query_6_5,connection)


C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [50]:
# результат запроса
df_6_5

avg
0  6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [51]:
# текст запроса
query_6_6 = f'''  
SELECT 
    v.experience,
    ROUND(AVG(COALESCE((v.salary_from + v.salary_to)/2, v.salary_to,v.salary_from))) average
FROM public.vacancies v
WHERE {DS_query}
    AND (v.salary_from is not null or v.salary_to is not null)
    AND (experience IS NOT NULL)
GROUP BY 1
'''
df_6_6 = pd.read_sql_query(query_6_6,connection)

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [52]:
# результат запроса
df_6_6

experience   average
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

# Выводы по предметному анализу:

На данном этапе мы исследовали все вакансии связанные с Data Science. 
* Для начала мы вычислили количество вакансий имеющиюх отношение к данным, их оказалось 1771. 
* Далее мы искали количество вакансий для начинающего дата-сайентиста, то-есть уровень junior, без опыта или в качестве стажировки - таких вакансий в нашей базе 51. Этот результат немного расстраивает, так как таких вакансий действительно не много и многие дата-сайенисты, с хорошей базой знаний попросту не могут найти работу из-за отсутствия опыта работы.  
* Так-же, мы решили узнать, сколько есть дата-сайенс вакансий с требоаниями к знанию SQL, таких мы нашли 201. 
* Вакансий с требованием к знанию языка Python оказалось немного больше, их в нашем дата-сете 351. Это меня удивило, так как я считал Python основным навыком любого дата-сайентиста, и был удивлен, когда оказалось, что не так то много вакансий его требуют. 
* Далее мы выявили сколько ключевых навыков в среднем указывают в вакансиях для дата-сайентистов и получили 6.41 - это довольно много, учитывая, что многие из этих навыков требуют многих лет обучения, однако специалист работающий с данными должен обладать большим кол-вом знаний, для эффективного развития предприятия.
* И напоследок мы вычислили среднюю зарплату для каждого типа требуемого опыта. Для начинающий дата-сайентистов, то-есть без опыта, средняя З/П составляет около 75 тыс, более опытным (опыт от 1 года и до 3 лет) средняя З/П уже составила около 140 тысяч, а уже у более продвинутыъ дата-сайентистов, с опытом от 3 до 6 лет, средняя зарплата превысила 240 тысяч рублей, что неудивительно, так как опыт чаще всего и влияет на уровень зарплаты.
 

***

# Общий вывод по проекту

Большинство выводов я уже сделал на каждом этапе работы над проектом, так что тут лиш кратко опишу уже сказанное:

* Вакансий у нас всего около 50 тысяч, публикованы они были большим количеством разных работодателей, коих насчитывается около 23 тысяч в нашем дата-сете. 
* Эти вакансии публикованы на обширной территории, в более чем 1300 регионов
* Вакансии разделены на почти 300 разных сфер деятельности 

## Далее я представляю несколько графиков с описанием:

1. Колиичество вакансий в разных городах

In [53]:

fig = px.bar(
    df_4_1,
    x='area',
    y='cnt',
    title='Количество вакансий в разных городах',
    width=800,
    height=600
)
fig.write_image('data/plot1.png')
#fig.show()

Как мы видим, в Москве наибольшее количество вакансий, что неудивительно, так как это самый большой город России и там в основном находяться все самые изветсные компании, которые предоставляют наибольшее количество рабочих мест. Практически в два раза меньше вакансий мы видим в городе-герое Санкт-Петерберг, ну и далее количество вакансий - как и ожидалось - все уменьшается и уменьшается. Последнее место занимает город Воронеж с чуть более чем одной тысячей вакансий. Однако это только самые большие города и развитые города. Я не стал включать те города, в которых количество вакансий менее одной тысячи.

2. Количество вакансий и опыт работы

In [54]:

fig = px.bar(
    df_4_5,
    x='experience',
    y='count',
    title='Количество вакансий и опыт работы',
    width=800,
    height=600
)
fig.write_image('data/plot2.png')
#fig.show()

В основном, мы имеем вакансии с требованием к опыту работы от 1 года и до 3 лет, это обуславливается тем, что компаниям не нужно очень много тим-лидов, сеньеров и так далее, так как им надо соответственно платить. Зачем нанимать много опытных сотрудников для задач, с которыми справятся и менее опытные, которым и платить можно будет меньше? По такой логике как мне кажется, компании и нанимают в основном не очень опытных сотрудников в большом количестве, для решения несложных задач, и несколько сеньеров и тим-лидов для каждого отдела, что-бы те проверяли работу первых.

3. Работадатели с наибольшим кол-вом вакансий

In [55]:

query = f'''
SELECT 
    e.name,
    COUNT(v.id)
FROM public.employers e
    JOIN vacancies v ON v.employer_id = e.id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 5
'''
df = pd.read_sql_query(query,connection)
fig = px.pie(
    df,
    values='count',
    names='name',
    title='Работадатели с наибольшим количеством вакансий',
    width=600,
    height=600
)
fig.write_image('data/plot3.png')
#fig.show()

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning:

pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy



Здесь длительное описание не требуется, так как результат вполне ожидаемый, мы видим Яндекс занимает более 50 процентов всего рынка, далее идут остальные крупные компании, такие как Ростелеком, Тинькофф, СБЕР и Газпром нефть, каждая из которых имеет от 9 до 13 процентов от рынка вакансий.

1. Распределение вакансий дата-саентистов в зависимости от опыта работы.

In [56]:
# salary distribution based on experience
query = f'''  
SELECT 
    v.experience,
    COALESCE((v.salary_from + v.salary_to)/2, v.salary_to,v.salary_from) salary
FROM public.vacancies v
WHERE {DS_query}
    AND (v.salary_from is not null or v.salary_to is not null)
    AND (experience IS NOT NULL)
'''
df = pd.read_sql_query(query,connection)
fig = px.box(
    df,
    x='salary',
    y='experience',
    title='Распределение зарплаты дата-сайентистов в зависимости от опыта',
    width=800,
    height=600
)
fig.write_image('data/plot4.png')
#fig.show()

C:\Users\German\AppData\Roaming\Python\Python310\site-packages\pandas\io\sql.py:762: UserWarning:

pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy



Тут мы видим подтверждение моих слов из графика под номером 2. - более опытные специалисты, так-же и более оплачеваемые, тут мы видим прямую взаимосвязь. 
С опытом от 3 до 6 лет, межквартильный размах заработной платы находиться между 200 и 300 тысячами рублей, в то время как с опытом от 1 года и до 3 лет цифры уже гораздо ниже - 175 тыс верхний квартиль и 95 тысяч нижний. Что касается совсем новичков без опыта, то их зп с основном находиться между 55 и 93 тысячами рублей.